ext/mysql2/client.c

```diff
@@ -504,26 +504,34 @@ static int opt_connect_attr_add_i(VALUE key, VALUE value, VALUE arg)
 }
 #endif
 
-static VALUE rb_mysql_connect(VALUE self, VALUE user, VALUE pass, VALUE host, VALUE port, VALUE database, VALUE socket, VALUE flags, VALUE conn_attrs) {
+static VALUE rb_mysql_connect(VALUE self, VALUE user, VALUE pass, VALUE host, VALUE port, VALUE database, VALUE socket, VALUE flags, VALUE conn_attrs, VALUE tls_sni_name) {
   struct nogvl_connect_args args;
   time_t start_time, end_time, elapsed_time, connect_timeout;
+  const char *sni_hostname;
   VALUE rv;
   GET_CLIENT(self);
 
-  args.host        = NIL_P(host)     ? NULL : StringValueCStr(host);
-  args.unix_socket = NIL_P(socket)   ? NULL : StringValueCStr(socket);
-  args.port        = NIL_P(port)     ? 0    : NUM2INT(port);
-  args.user        = NIL_P(user)     ? NULL : StringValueCStr(user);
-  args.passwd      = NIL_P(pass)     ? NULL : StringValueCStr(pass);
-  args.db          = NIL_P(database) ? NULL : StringValueCStr(database);
-  args.mysql       = wrapper->client;
-  args.client_flag = NUM2ULONG(flags);
+  args.host         = NIL_P(host)     ? NULL : StringValueCStr(host);
+  args.unix_socket  = NIL_P(socket)   ? NULL : StringValueCStr(socket);
+  args.port         = NIL_P(port)     ? 0    : NUM2INT(port);
+  args.user         = NIL_P(user)     ? NULL : StringValueCStr(user);
+  args.passwd       = NIL_P(pass)     ? NULL : StringValueCStr(pass);
+  args.db           = NIL_P(database) ? NULL : StringValueCStr(database);
+  args.mysql        = wrapper->client;
+  args.client_flag   = NUM2ULONG(flags);
+
+  sni_hostname      = NIL_P(tls_sni_name) ? NULL : StringValueCStr(tls_sni_name);
 
 #ifdef CLIENT_CONNECT_ATTRS
   mysql_options(wrapper->client, MYSQL_OPT_CONNECT_ATTR_RESET, 0);
   rb_hash_foreach(conn_attrs, opt_connect_attr_add_i, (VALUE)wrapper);
 #endif
 
+  if(sni_hostname != NULL) {
+    /* Set the TLS SNI name if provided */
+    mysql_options(wrapper->client, MYSQL_OPT_TLS_SNI_SERVERNAME, sni_hostname);
```

Comment:
I found that this wasn't added until MySQL 8.1: 
https://dev.mysql.com/doc/relnotes/mysql/8.1/en/news-8-1-0.html

However MySQL 8.0 is still in support window